test/test_inverse.py

```diff
@@ -704,40 +775,72 @@ def test_KFAC_inverse_save_and_load_state_dict():
 
     # check that the two inverse KFACs are equal
     compare_state_dicts(inv_kfac.state_dict(), inv_kfac_new.state_dict())
-    test_vec = torch.rand(inv_kfac.shape[1])
-    report_nonclose(inv_kfac @ test_vec, inv_kfac_new @ test_vec)
+    report_nonclose(test_mvp, inv_kfac_new @ test_vec)
 
 
-def test_KFAC_inverse_from_state_dict():
+@mark.parametrize("use_exact_damping", [True, False], ids=["exact_damping", ""])
+@mark.parametrize("use_heuristic_damping", [True, False], ids=["heuristic_damping", ""])
+@mark.parametrize("cache", [True, False], ids=["cached", "uncached"])
+@mark.parametrize(
+    "exclude", [None, "weight", "bias"], ids=["all", "no_weights", "no_biases"]
+)
+@mark.parametrize(
+    "separate_weight_and_bias", [True, False], ids=["separate_bias", "joint_bias"]
+)
+def test_KFAC_inverse_from_state_dict(
+    use_exact_damping: bool,
+    use_heuristic_damping: bool,
+    cache: bool,
+    exclude: str,
+    separate_weight_and_bias: bool,
+):
     """Test that KFACInverseLinearOperator can be created from state dict."""
     torch.manual_seed(0)
-    batch_size, D_in, D_out = 4, 3, 2
+    batch_size, D_in, D_hidden, D_out = 4, 3, 5, 2
     X = torch.rand(batch_size, D_in)
     y = torch.rand(batch_size, D_out)
-    model = torch.nn.Linear(D_in, D_out)
+    model = torch.nn.Sequential(
+        torch.nn.Linear(D_in, D_hidden),
+        torch.nn.ReLU(),
+        torch.nn.Linear(D_hidden, D_hidden, bias=False),
+        torch.nn.ReLU(),
+        torch.nn.Linear(D_hidden, D_out),
+    )
 
     params = list(model.parameters())
+    if exclude is not None:
+        names = {p.data_ptr(): name for name, p in model.named_parameters()}
+        params = [p for p in params if exclude not in names[p.data_ptr()]]
+
     # create and compute KFAC
     kfac = KFACLinearOperator(
         model,
         MSELoss(reduction="sum"),
         params,
         [(X, y)],
+        separate_weight_and_bias=separate_weight_and_bias,
     )
 
     # create inverse KFAC and save state dict
+    kwargs = {
+        "use_exact_damping": use_exact_damping,
+        "use_heuristic_damping": use_heuristic_damping,
+    }
+    if use_exact_damping and use_heuristic_damping:
+        return
     inv_kfac = KFACInverseLinearOperator(
-        kfac, damping=1e-2, use_heuristic_damping=True, retry_double_precision=False
+        kfac, damping=1e-2, retry_double_precision=False, cache=cache, **kwargs
     )
+    test_vec = torch.rand(kfac.shape[1])
+    test_mvp = inv_kfac @ test_vec  # triggers inverse computation and maybe caching
```

Comment:
Question: Why does `@` work on a PyTorch tensor here? Shouldn't `inv_kfac` process `numpy` arrays, at least before we merge #149?

Reply:
It should work with PyTorch tensors as long as they are on CPU, because of [this](https://github.com/scipy/scipy/blob/v1.14.1/scipy/sparse/linalg/_interface.py#L330-L331). But you're right that this goes against our type hints and only works on CPU, so I'm happy to change to a numpy array here.